# 0.1 - Data Collection-1

- SpaceX REST API - launch data, rocket used, payload delivered, launch specs, landing specs, landing outcome.

In [1]:
%load_ext autoreload
%autoreload 2

import pandas as pd
import numpy as np
import requests
import json

# Pandas setting to completely show table
pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth', None)

url = "https://api.spacexdata.com/v4/launches/past"
response = requests.get(url)
try:
    response.raise_for_status()
except requests.exceptions.HTTPError as e:
    print(e)

525 Server Error: <none> for url: https://api.spacexdata.com/v4/launches/past


It's a Cloudflare error...

The server of SpaceX REST API is down, so I had to look for other sources. But the static JSON provided by the course for project use still proved to be useful for retrieving corse data -- `gridfins_True`, `gridfins_False`, `legs_True`, `legs_False`. Furthermore, with the static JSON, I was still able to follow the time period of launches that is considered in the project.

## Data Collection from other sources

![falcon-9-components](../references/falcon-9-scale-and-transportation-infographic-v0-AJTKrBYPWvz-es3BQYfmOhiqpFl72jFH-IUBGRT43LM.webp)

The following is the list of features that are needed:
- **Booster Version** - `str`; The rocket engine used for the launch, e.g., Falcon Heavy, Falcon 9.
- **Payload Mass** - `float`; The total mass (in kg) of cargo carried by the spacecraft during the launch, e.g., 800 kg
- **Orbit** - `str`; The type of orbit in which the spacecraft is placed upon launch, e.g., Sun-synchronous orbit (SSO), Low Earth orbit (LEO).
- **LaunchSite** - `str`; name of the launch site, e.g., VAFB SLC 4E
- **Outcome** - `str`; Concatenation of landing success (`bool`) and landing type ('str'); e.g., 'True ASDS'. ASDS means autonomous spaceport drone ship. 
- **Flights** - `int`; number of flights with the core used in a launch
- **GridFins** - `bool`; whether or not the spacecraft has grid fins.
- **Reused** - `bool`; whether or not the spacecraft is reused for other flights.
- **Legs** - `bool`; whether or not the spacecraft has landing legs.
- **LandingPad** - `str`; name of the landing pad, e.g., LZ-2
- **Block** - `int`; version of the rocket, e.g., Falcon 9 block 5
- **ReusedCount** - `int`; number of times the rocket is reused.
- **Serial** - `str`; identification of the booster used, e.g., B1051
- **Longitude** - `float`; longitude of the launchpad.
- **Latitude** - `float`; latitude of the launchpad.

Features needed, the reference where data is retrieved, and the corresponding item in the reference.

| Feature | Reference (API Endpoint / JSON / TSV) | Item |
| --- | --- | --- |
| `BoosterVersion` | [LL2 API previous launches](https://ll.thespacedevs.com/docs/#/launches/launches_previous_list) | `rocket.configuration.name` |
| `Orbit` | [LL2 API previous launches](https://ll.thespacedevs.com/docs/#/launches/launches_previous_list) | `mission.orbit.name`|
| `Payload mass` | [General Catalog of Artificial Space Objects](https://planet4589.org/space/gcat/web/launch/lcols.html) | `OrbPay`|
| `LaunchSite` | [LL2 API previous launches](https://ll.thespacedevs.com/docs/#/launches/launches_previous_list) | `pad.name` |
| `Outcome` | [LL2 API previous launches](https://ll.thespacedevs.com/docs/#/launches/launches_previous_list) | `rocket.launcher_stage.landing.success` & `rocket.launcher_stage.landing.type.abbrev`|
| `Flights` | [LL2 API previous launches](https://ll.thespacedevs.com/docs/#/launches/launches_previous_list) | `rocket.launcher_stage.launcher.flights` |
| `Gridfins` | [Static JSON provided by the course for project use](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/API_call_spacex_api.json) | `cores.gridfins` |
| `Reused` | [LL2 API previous launches](https://ll.thespacedevs.com/docs/#/launches/launches_previous_list) | `rocket.launcher_stage.launcher.reused` | 
| `Legs` | [Static JSON provided by the course for project use](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/API_call_spacex_api.json) | `cores.legs` |
| `LandingPad` | [LL2 API previous launches](https://ll.thespacedevs.com/docs/#/launches/launches_previous_list) | `rocket.launcher_stage.landing.location.abbrev` |
| `Block` | [LL2 API previous launches](https://ll.thespacedevs.com/docs/#/launches/launches_previous_list) | `rocket.configuration.variant` |
| `ReusedCount` | [LL2 API previous launches](https://ll.thespacedevs.com/docs/#/launches/launches_previous_list) | `(rocket.launcher_stage.launcher_flight_number) - 1` |
| `Serial` | [LL2 API previous launches](https://ll.thespacedevs.com/docs/#/launches/launches_previous_list) | `rocket.launcher_stage.launcher.serial_number` |
| `Longitude` | [LL2 API previous launches](https://ll.thespacedevs.com/docs/#/launches/launches_previous_list) | `pad.longitude` |
| `Latitude` | [LL2 API previous launches](https://ll.thespacedevs.com/docs/#/launches/launches_previous_list) | `pad.latitude` |

### Data from Launch Library 2 (LL2) API

Since these are API calls, a script was made to retrieve them, combine  and to cache each call as JSON files saved in `data/raw`

In [2]:
from pathlib import Path
import time

import requests


def get_ll2_launches(offset):
    ll2_api = "https://ll.thespacedevs.com/2.3.0/launches/previous/"
    query_params = dict(
        mode="detailed", limit=100, rocket__configuration__name="Falcon 9", offset=offset
    )

    for attempt in range(3):
        try:
            response = requests.get(ll2_api, params=query_params)
            if response.status_code == 504:
                print("504 Server Error. Retrying in {} seconds...".format(2 * (attempt + 1)))
                time.sleep(2 * (attempt + 1))
                continue
            response.raise_for_status()
            return response.json()

        except requests.exceptions.HTTPError as e:
            print("Error: {}".format(e))
            return None

    print("Failed after 3 attempts (504 Server Error)")
    return None

def download_all_ll2_launches():
    file_paths = []
    for offset in range(0, 700, 100):
        parent_dir = Path.cwd().parent
        file_dir = parent_dir / "data" / "raw"
        file_dir.mkdir(parents=True, exist_ok=True)
        file_name = "ll2-api-2.3.0-launches-previous-{}.json".format(offset)
        file_path = file_dir / file_name
        file_paths.append(file_path)
        if file_path.is_file():
            print("{} already exists.".format(file_name))
            continue
        try:
            data = get_ll2_launches(offset)
            if not isinstance(data, dict):
                raise TypeError("Expected dict, got {}".format(type(data)))
            with open(file_path, "w") as f:
                json.dump(data, f)
        except Exception as e:
            print("Error: {}".format(e))
            break
    return file_paths

In [3]:
download_all_ll2_launches()

ll2-api-2.3.0-launches-previous-0.json already exists.
ll2-api-2.3.0-launches-previous-100.json already exists.
ll2-api-2.3.0-launches-previous-200.json already exists.
ll2-api-2.3.0-launches-previous-300.json already exists.
ll2-api-2.3.0-launches-previous-400.json already exists.
ll2-api-2.3.0-launches-previous-500.json already exists.
ll2-api-2.3.0-launches-previous-600.json already exists.


[PosixPath('/Users/jelo/spacey_falcon_9_project/data/raw/ll2-api-2.3.0-launches-previous-0.json'),
 PosixPath('/Users/jelo/spacey_falcon_9_project/data/raw/ll2-api-2.3.0-launches-previous-100.json'),
 PosixPath('/Users/jelo/spacey_falcon_9_project/data/raw/ll2-api-2.3.0-launches-previous-200.json'),
 PosixPath('/Users/jelo/spacey_falcon_9_project/data/raw/ll2-api-2.3.0-launches-previous-300.json'),
 PosixPath('/Users/jelo/spacey_falcon_9_project/data/raw/ll2-api-2.3.0-launches-previous-400.json'),
 PosixPath('/Users/jelo/spacey_falcon_9_project/data/raw/ll2-api-2.3.0-launches-previous-500.json'),
 PosixPath('/Users/jelo/spacey_falcon_9_project/data/raw/ll2-api-2.3.0-launches-previous-600.json')]

### Data provided by course

The launch data provided by the course and as well as data from Jonathan McDowell's GCAT are considered static files and were downloaded with the following utility function I wrote:

In [4]:
def download_launch_data_static(url, file_name):
    parent_dir = Path.cwd().parent
    file_dir = parent_dir / "data" / "raw"
    file_dir.mkdir(parents=True, exist_ok=True)
    file_path = file_dir / file_name

    if file_path.is_file():
        print("{} already exists".format(file_name))
        return file_path
    
    response = requests.get(url)
    
    try:
        response.raise_for_status()
        
    except requests.exceptions.HTTPError as e:
        print("Error: {}".format(e))
        return None

    try:
        with open(file_path, "wb") as f:
            f.write(response.content)
        return file_path
    except Exception as e:
        print("Error: {}".format(e))
        return None

In [5]:
url_ibm='https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/API_call_spacex_api.json'
file_name_ibm = 'ibm-ds-capstone-launch-data.json'
download_launch_data_static(url_ibm, file_name_ibm)

ibm-ds-capstone-launch-data.json already exists


PosixPath('/Users/jelo/spacey_falcon_9_project/data/raw/ibm-ds-capstone-launch-data.json')

### Data from Jonathan McDowell's General Catalogue of Artificial Space Objects

Payload data from the LL2 API is new and only contains data for launches in 2026. Hence, I relied on Jonathan McDowell's GCAT for payload data. Data from this source was available, and hence downloaded, as a `.tsv` file.

In [6]:
url_gcat='https://planet4589.org/space/gcat/tsv/launch/Falcon9.tsv'
file_name_gcat = 'mcdowell-gcat-launch-data.tsv'
gcat_path = download_launch_data_static(url_gcat, file_name_gcat)

mcdowell-gcat-launch-data.tsv already exists


## Merging the data

After collecting the raw data, I have merged the three sources into one `csv` file and saved it in `data/interim` for later processing.

### Launch Library 2

In [7]:
par_dir = Path.cwd().parent

file_path = par_dir / 'data/raw/ll2-api-2.3.0-launches-previous-500.json'
with open(file_path, 'r') as f:
    data_json = json.load(f)['results']

In [8]:
def get_path(d, path):
    for key in path.split('.'):
        if isinstance(d, list):
            d = d[0]
        d = d.get(key) if isinstance(d, dict) else None
    return d 

In [9]:
path = 'rocket.configuration.name'
booster_version = []
for launch in data_json:
    booster_version.append(get_path(launch, path))
pd.Series(booster_version)

0     Falcon 9
1     Falcon 9
2     Falcon 9
3     Falcon 9
4     Falcon 9
        ...   
95    Falcon 9
96    Falcon 9
97    Falcon 9
98    Falcon 9
99    Falcon 9
Length: 100, dtype: str

In [10]:
json_paths = dict(
    launch_designator = 'launch_designator',
    window_start = 'window_start',
    booster_version = 'rocket.configuration.name',
    orbit = 'mission.orbit.name',
    launch_site = 'pad.name',
    landing_success = 'rocket.launcher_stage.landing.success',
    landing_type = 'rocket.launcher_stage.landing.type.abbrev',
    flights = 'rocket.launcher_stage.launcher.flights',
    reused = 'rocket.launcher_stage.reused',
    landing_pad = 'rocket.launcher_stage.landing.landing_location.abbrev',
    block = 'rocket.configuration.variant',
    flight_number = 'rocket.launcher_stage.launcher_flight_number',
    serial = 'rocket.launcher_stage.launcher.serial_number',
    longitude = 'pad.longitude',
    latitude = 'pad.latitude'
)

launch_dict = {}
for col, path in json_paths.items():
    col_values = []
    for launch in data_json:
        col_values.append(get_path(launch, path))
    launch_dict.update({col: col_values})

In [11]:
path = json_paths['flight_number']
keys = path.split('.')
print('{}\n'.format(keys))

launch = data_json[4]

key = keys[0]
d = launch.get(key)
print('Key 0:\n{}\n'.format(d.keys()))

key = keys[1]
d = d.get(key)
d = d[0]
print('Key 1:\n{}\n'.format(d.keys()))

key = keys[2]
d = d.get('launcher_flight_number')
print(d)
#print('Key 2:\n{}\n'.format(d.keys()))

#key = keys[3]
#d = d.get('landing_location')
#print('Key 3:\n{}\n'.format(d.keys()))

['rocket', 'launcher_stage', 'launcher_flight_number']

Key 0:
dict_keys(['id', 'configuration', 'launcher_stage', 'spacecraft_stage', 'payloads'])

Key 1:
dict_keys(['id', 'type', 'reused', 'launcher_flight_number', 'launcher', 'previous_flight_date', 'turn_around_time', 'landing', 'previous_flight'])

3


In [12]:
print(launch.keys())

dict_keys(['id', 'url', 'name', 'response_mode', 'slug', 'launch_designator', 'status', 'last_updated', 'net', 'net_precision', 'window_end', 'window_start', 'image', 'infographic', 'probability', 'weather_concerns', 'failreason', 'hashtag', 'launch_service_provider', 'rocket', 'mission', 'pad', 'webcast_live', 'program', 'orbital_launch_attempt_count', 'location_launch_attempt_count', 'pad_launch_attempt_count', 'agency_launch_attempt_count', 'orbital_launch_attempt_count_year', 'location_launch_attempt_count_year', 'pad_launch_attempt_count_year', 'agency_launch_attempt_count_year', 'flightclub_url', 'updates', 'info_urls', 'vid_urls', 'timeline', 'pad_turnaround', 'mission_patches'])


In [13]:
data_df = pd.DataFrame(launch_dict)

In [14]:
data_df.head()

,launch_designator,window_start,booster_version,orbit,launch_site,landing_success,landing_type,flights,reused,landing_pad,block,flight_number,serial,longitude,latitude
0,2022-077,2022-07-11T01:39:40Z,Falcon 9,Low Earth Orbit,Space Launch Complex 4E,True,ASDS,33,True,OCISLY,Block 5,6,B1063,-120.611000,34.632000
1,2022-076,2022-07-07T13:11:00Z,Falcon 9,Low Earth Orbit,Space Launch Complex 40,True,ASDS,19,True,JRTI,Block 5,13,B1058,-80.577357,28.561941
2,2022-071,2022-06-29T21:04:00Z,Falcon 9,Geostationary Transfer Orbit,Space Launch Complex 40,True,ASDS,21,True,ASOG,Block 5,2,B1073,-80.577357,28.561941
3,2022-064,2022-06-19T04:27:36Z,Falcon 9,Low Earth Orbit,Space Launch Complex 40,True,ASDS,23,True,JRTI,Block 5,9,B1061,-80.577357,28.561941
4,2022-063,2022-06-18T14:19:52Z,Falcon 9,Sun-Synchronous Orbit,Space Launch Complex 4E,True,RTLS,35,True,LZ-4,Block 5,3,B1071,-120.611000,34.632000


In [15]:
data_df['reused_count'] = data_df['flight_number'] - 1
data_df = data_df.drop('flight_number', axis=1)

In [16]:
data_df['launch_date'] = pd.to_datetime(data_df['window_start']).dt.date

In [17]:
data_df['launch_date'].head()

0    2022-07-11
1    2022-07-07
2    2022-06-29
3    2022-06-19
4    2022-06-18
Name: launch_date, dtype: object

In [18]:
data_df.dtypes

launch_designator        str
window_start             str
booster_version          str
orbit                    str
launch_site              str
landing_success       object
landing_type             str
flights                int64
reused                  bool
landing_pad              str
block                    str
serial                   str
longitude            float64
latitude             float64
reused_count           int64
launch_date           object
dtype: object

In [19]:
data_df['outcome'] = data_df['landing_success'].astype(str) + ' ' + data_df['landing_type']
data_df = data_df.drop(['landing_success', 'landing_type'], axis=1)

In [20]:
data_df['launch_designator'].isnull().value_counts()

launch_designator
False    99
True      1
Name: count, dtype: int64

#### Script for querying data from the LL2 raw data

Based on the above experimentation, I created a script to query relevant data from the raw LL2 data. But before that, it makes sense to add a script to merge the paginated files into one json file for easier transformation.

In [21]:
def merge_ll2_launch_data(json_paths: list[Path], merged_path: None|Path = None) -> Path:
    '''
    Function to merge LL2 launches that were extracted and downloaded with the download_all_ll2_launches() function

    :param json_paths: list[Path]; file paths to the paginated files. Particularly, the returned output of the
                       download_all_ll2_launches() function
    :param merged_path: None or Path; path to the merged files. If not specified, the merged file will be saved in
                        the current working directory.
    '''

    merged = []
    for path in json_paths:
        with open(path, 'r') as f:
            launch_page = json.load(f)
        merged.extend(launch_page['results'])

    if merged_path is None:
        merged_path = Path.cwd() / 'll2-api-2.3.0-launches-previous-merged.json'

    with open(merged_path, 'w') as f:
        json.dump(merged, f)

    return merged_path

In [22]:
json_paths = download_all_ll2_launches()
merged_path = Path.cwd().parent / 'data' / 'interim' / 'll2-api-2.3.0-launches-previous-merged.json'
merge_ll2_launch_data(json_paths, merged_path)

ll2-api-2.3.0-launches-previous-0.json already exists.
ll2-api-2.3.0-launches-previous-100.json already exists.
ll2-api-2.3.0-launches-previous-200.json already exists.
ll2-api-2.3.0-launches-previous-300.json already exists.
ll2-api-2.3.0-launches-previous-400.json already exists.
ll2-api-2.3.0-launches-previous-500.json already exists.
ll2-api-2.3.0-launches-previous-600.json already exists.


PosixPath('/Users/jelo/spacey_falcon_9_project/data/interim/ll2-api-2.3.0-launches-previous-merged.json')

Awesome!

In [23]:
from typing import Any

def get_path(d: dict, path: str) -> None|Any:
    '''
    A utility function for quickly accessing items in a nested dictionary
    from a specified string of keys joined by dots '.'

    :param d: dict; nested dictionary containing items to be accessed
    :param path: str; string of keys joined by dots '.'
    :return: None or Any; the item to be accessed, whatever its type 
    '''
    for key in path.split('.'):
        if isinstance(d, list):
            d = d[0]
        d = d.get(key) if isinstance(d, dict) else None
    return d

def transform_ll2_launches(json_path: str|Path) -> pd.DataFrame:
    '''
    A function to transform LL2 raw data given as a json file path
    
    :param json_path: str or Path; local path to the json file containing raw ll2 launches data
    :return: pd.DataFrame; transformed LL2 launch data containing relevant columns for analysis, but requires further wrangling
    '''
    with open(json_path, 'r') as f:
        data_json = json.load(f)

    json_keys = dict(
        launch_designator = 'launch_designator',
        net = 'net',
        booster_version = 'rocket.configuration.name',
        orbit = 'mission.orbit.name',
        launch_site = 'pad.name',
        landing_success = 'rocket.launcher_stage.landing.success',
        landing_type = 'rocket.launcher_stage.landing.type.abbrev',
        flights = 'rocket.launcher_stage.launcher.flights',
        reused = 'rocket.launcher_stage.reused',
        landing_pad = 'rocket.launcher_stage.landing.landing_location.abbrev',
        block = 'rocket.configuration.variant',
        flight_number = 'rocket.launcher_stage.launcher_flight_number',
        serial = 'rocket.launcher_stage.launcher.serial_number',
        longitude = 'pad.longitude',
        latitude = 'pad.latitude'
    )
    
    launch_dict = {}
    for col, path in json_keys.items():
        col_values = []
        for launch in data_json:
            col_values.append(get_path(launch, path))
        launch_dict.update({col: col_values})

    data_df = pd.DataFrame(launch_dict)

    # Derived column - reused_count
    data_df['reused_count'] = data_df['flight_number'] - 1
    data_df = data_df.drop('flight_number', axis=1)

    # Derived column - launch_date
    data_df['launch_date'] = pd.to_datetime(data_df['net']).dt.date
    data_df = data_df.drop('net', axis=1)

    # Derived column - outcome
    data_df['outcome'] = data_df['landing_success'].astype(str) + ' ' + data_df['landing_type']
    data_df = data_df.drop(['landing_success', 'landing_type'], axis=1)

    # Drop missing values in launch designator and outcome
    data_df = data_df.dropna(subset=['launch_designator', 'outcome'])

    return data_df

In [24]:
ll2_df = transform_ll2_launches(merged_path)

In [25]:
ll2_df.info()

<class 'pandas.DataFrame'>
Index: 623 entries, 6 to 663
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   launch_designator  623 non-null    str    
 1   booster_version    623 non-null    str    
 2   orbit              622 non-null    str    
 3   launch_site        623 non-null    str    
 4   flights            623 non-null    int64  
 5   reused             623 non-null    bool   
 6   landing_pad        623 non-null    str    
 7   block              623 non-null    str    
 8   serial             623 non-null    str    
 9   longitude          623 non-null    float64
 10  latitude           623 non-null    float64
 11  reused_count       623 non-null    int64  
 12  launch_date        623 non-null    object 
 13  outcome            623 non-null    str    
dtypes: bool(1), float64(2), int64(2), object(1), str(8)
memory usage: 68.7+ KB


### General Catalogue of Artificial Space Objects

In [26]:
gcat_path = par_dir / 'data/raw/mcdowell-gcat-launch-data.tsv'
gcat_data = pd.read_csv(gcat_path, sep='\t', skiprows = (lambda x: x in [1]))
gcat_data.head()

,#Launch_Tag,Launch_JD,Launch_Date,LV_Type,Variant,Flight_ID,Flight,Mission,FlightCode,Platform,Launch_Site,Launch_Pad,Ascent_Site,Ascent_Pad,Apogee,Apoflag,Range,RangeFlag,Dest,OrbPay,Agency,LaunchCode,FailCode,Group,Category,LTCite,Cite,Notes
0,2010-026,2455352.28,2010 Jun 4 1845:00,Falcon 9,-,001/B0003? F001,Dragon Qual Unit,Dragon Spacecraft,-,-,CC,LC40,-,-,276,,-,,-,6.000,SPX,OS,-,CX,Sat LEO O,Wire,-,-
1,2010-066,2455539.15,2010 Dec 8 1543,Falcon 9,-,002/B0004? F002,Dragon C1,Dragon C101,-,-,CC,LC40,-,-,11084,,-,,-,8.774,SPX,OS,-,CX,Sat LEO O,Wire,-,-
2,2012-027,2456069.82,2012 May 22 0744:38,Falcon 9,-,003/B0005? F003,Dragon C2,Dragon C102,-,-,CC,LC40,-,-,453,,-,,-,7.379,SPX,OS,-,CX,Sat ISS O,Wire,-,-
3,2012-A10,2456191.50,2012 Sep 21,Grasshopper,-,B0002?,-,-,-,-,SPXMCG,-,-,-,0,,-,,-,0.000,SPX,AS,-,-,Test,spacex.com,-,2m hop
4,2012-054,2456208.52,2012 Oct 8 0035:07,Falcon 9,-,004/B0006? F004,Dragon CRS-1,Dragon C103,-,-,CC,LC40,-,-,426,,-,,-,8.249,SPX,OS85,S2R,CX,Sat ISS O,Wire,-,S1 engine out stopped S2R


In [27]:
import re
pattern = r'^\d{4}\s+\w{3}\s+\d{1,2}'
format_date = re.compile(pattern)

launch_date = gcat_data['Launch_Date'].map(lambda x: format_date.match(x).group())
gcat_data['Launch_Date'] = pd.to_datetime(launch_date)

In [28]:
gcat_data = gcat_data[['Launch_Date', 'OrbPay']]

In [29]:
gcat_data.head()

,Launch_Date,OrbPay
0,2010-06-04,6.000
1,2010-12-08,8.774
2,2012-05-22,7.379
3,2012-09-21,0.000
4,2012-10-08,8.249


#### Script for querying data

A script is then made for querying from the cached file containing data from GCAT.

In [30]:
def transform_gcat_data(gcat_path: str|Path) -> pd.DataFrame:
    '''
    Function to transform GCAT raw data from the specified file path

    :param gcat_path: str or Path; file path to the GCAT raw data.
    :return: pd.DataFrame; dataframe containing relevant data from GCAT
    '''

    gcat_df = pd.read_csv(gcat_path, sep='\t', skiprows = (lambda x: x in [1]))

    pattern = r'^\d{4}\s+\w{3}\s+\d{1,2}'
    format_date = re.compile(pattern)
    launch_date = gcat_df['Launch_Date'].map(lambda x: format_date.match(x).group())
    gcat_df['launch_date'] = pd.to_datetime(launch_date).dt.date

    gcat_df = gcat_df[['#Launch_Tag', 'launch_date', 'OrbPay']]
    gcat_df = gcat_df.rename(columns={'#Launch_Tag': 'launch_designator'}) #rename '#Launch_Tag' to launch designator
    gcat_df['launch_designator'] = gcat_df['launch_designator'].str.strip() #remove trailing white spaces from launch_designator column
    
    return gcat_df

In [31]:
gcat_df = transform_gcat_data(gcat_path)

### Course-provided Static JSON

In [32]:
json_path = download_launch_data_static(url_ibm, file_name_ibm)

ibm-ds-capstone-launch-data.json already exists


In [33]:
with open(json_path, 'r') as f:
    json_data = json.load(f)

json_keys = {
    'date_utc': 'date_utc',
    'gridfins': 'cores.gridfins',
    'legs': 'cores.legs'
}

course_dict = {}
for col, key in json_keys.items():
    cols = []
    for launch in json_data:
        cols.append(get_path(launch, key))
    course_dict.update({col: cols})

course_df = pd.DataFrame(course_dict)

In [34]:
course_df

,date_utc,gridfins,legs
0,2006-03-24T22:30:00.000Z,False,False
1,2007-03-21T01:10:00.000Z,False,False
2,2008-08-03T03:34:00.000Z,False,False
3,2008-09-28T23:15:00.000Z,False,False
4,2009-07-13T03:35:00.000Z,False,False
...,...,...,...
102,2020-10-06T11:29:00.000Z,True,True
103,2020-10-18T12:25:00.000Z,True,True
104,2020-10-24T15:31:00.000Z,True,True
105,2020-11-05T23:24:00.000Z,True,True


In [35]:
course_df['launch_date'] = pd.to_datetime(course_df['date_utc']).dt.date
course_df = course_df.drop('date_utc', axis=1)

In [36]:
course_df

,gridfins,legs,launch_date
0,False,False,2006-03-24
1,False,False,2007-03-21
2,False,False,2008-08-03
3,False,False,2008-09-28
4,False,False,2009-07-13
...,...,...,...
102,True,True,2020-10-06
103,True,True,2020-10-18
104,True,True,2020-10-24
105,True,True,2020-11-05


In [37]:
def transform_course_data(course_path: str|Path) -> pd.DataFrame:
    '''
    Function to transform course-provided static JSON raw data from the specified path

    :param course_path: str or Path; file path to the json raw data
    :return: pd.DataFrame; containing relevant data from the course-provided data
    '''

    with open(course_path, 'r') as f:
        json_data = json.load(f)
    
    json_keys = {
        'date_utc': 'date_utc',
        'gridfins': 'cores.gridfins',
        'legs': 'cores.legs'
    }
    
    course_dict = {}
    for col, key in json_keys.items():
        cols = []
        for launch in json_data:
            cols.append(get_path(launch, key))
        course_dict.update({col: cols})
    
    course_df = pd.DataFrame(course_dict)

    course_df['launch_date'] = pd.to_datetime(course_df['date_utc']).dt.date
    course_df = course_df.drop('date_utc', axis=1)
    
    return course_df

In [38]:
course_df = transform_course_data(json_path)

### The Merge

Now that data from each source have been transformed, it is now time to join the data into one pandas DataFrame.

In [39]:
print('ll2 df: {}\ngcat df: {}\ncourse df: {}'.format(ll2_df.shape, gcat_df.shape, course_df.shape))

ll2 df: (623, 14)
gcat df: (689, 3)
course df: (107, 3)


Based on the above shapes, the final merged dataframe should contain at most 107 samples -- so only a small portion of data was preserved from gcat and ll2 :( ...

In [41]:
import datetime

gcat_df[gcat_df['launch_date'].map(lambda x: x == datetime.date(2022, 12, 16))]

,launch_designator,launch_date,OrbPay
207,2022-173,2022-12-16,2.2
208,2022-174,2022-12-16,3.4


In [42]:
ll2_df[ll2_df['launch_date'].map(lambda x: x == datetime.date(2026, 4, 11))]

,launch_designator,booster_version,orbit,launch_site,flights,reused,landing_pad,block,serial,longitude,latitude,reused_count,launch_date,outcome
42,2026-079,Falcon 9,Low Earth Orbit,Space Launch Complex 40,7,True,LZ-40,Block 5,B1094,-80.577357,28.561941,6,2026-04-11,True RTLS
43,2026-077,Falcon 9,Low Earth Orbit,Space Launch Complex 4E,33,True,OCISLY,Block 5,B1063,-120.611000,34.632000,31,2026-04-11,True ASDS


From the above code, it can be seen that there may be more than one launch in a day. Hence, a one-to-one inner join on launch dates cannot be done. However, 'launch_designator' on `ll2_df` and '#Launch_Tag' on `gcat_df` can be used to join the ll2 launches and gcat launches, but only on these two dataframes. I turned back to launch date when merging the course data.

In [43]:
merged_df = ll2_df.merge(gcat_df, on='launch_designator', how='inner', validate='1:1')
merged_df.head()

,launch_designator,booster_version,orbit,launch_site,flights,reused,landing_pad,block,serial,longitude,latitude,reused_count,launch_date_x,outcome,launch_date_y,OrbPay
0,2026-157,Falcon 9,Low Earth Orbit,Space Launch Complex 40,36,True,ASOG,Block 5,B1067,-80.577357,28.561941,35,2026-07-09,True ASDS,2026-07-09,16.1
1,2026-156,Falcon 9,Sun-Synchronous Orbit,Space Launch Complex 4E,11,True,OCISLY,Block 5,B1097,-120.611000,34.632000,10,2026-07-07,True ASDS,2026-07-07,5.0
2,2026-154,Falcon 9,Low Earth Orbit,Space Launch Complex 40,13,True,ASOG,Block 5,B1090,-80.577357,28.561941,12,2026-07-05,True ASDS,2026-07-05,16.1
3,2026-150,Falcon 9,Low Earth Orbit,Space Launch Complex 4E,7,True,OCISLY,Block 5,B1100,-120.611000,34.632000,6,2026-07-02,True ASDS,2026-07-02,15.5
4,2026-148,Falcon 9,Geostationary Transfer Orbit,Space Launch Complex 40,17,True,ASOG,Block 5,B1085,-80.577357,28.561941,16,2026-06-29,True ASDS,2026-06-29,7.0


Check whether `launch_date_x` and `launch_date_y` columns are the same, replace them with a single `launch_date` column if they are.

In [44]:
mask = (merged_df['launch_date_x'] == merged_df['launch_date_y'])
mask.all()

np.True_

In [45]:
merged_df['launch_date'] = merged_df['launch_date_x']
merged_df = merged_df.drop(['launch_date_x', 'launch_date_y'], axis=1)

The next order of business is to combine course data and the merged df. Problem with that is there are duplicate dates, meaning there were more than one launches that occured in the same day.

In [46]:
merged_df[merged_df['launch_date'].map(lambda x: x == datetime.date(2026, 4, 11))]

,launch_designator,booster_version,orbit,launch_site,flights,reused,landing_pad,block,serial,longitude,latitude,reused_count,outcome,OrbPay,launch_date
36,2026-079,Falcon 9,Low Earth Orbit,Space Launch Complex 40,7,True,LZ-40,Block 5,B1094,-80.577357,28.561941,6,True RTLS,12.0,2026-04-11
37,2026-077,Falcon 9,Low Earth Orbit,Space Launch Complex 4E,33,True,OCISLY,Block 5,B1063,-120.611000,34.632000,31,True ASDS,15.5,2026-04-11


However, for the launches in the course data, there were no launches with overlapping dates. The range of dates for the course data is from 2006 to 2020.

In [47]:
course_df['launch_date'].unique().shape[0] == course_df.shape[0]

True

On the other hand, the range of dates for the merged df corresponding to launches with overlapping dates is from 2022 to 2026

In [48]:
counts = merged_df['launch_date'].value_counts()
duplicate_dates = merged_df[merged_df['launch_date'].map(lambda x: x in counts[counts>1].index)]
duplicate_dates

,launch_designator,booster_version,orbit,launch_site,flights,reused,landing_pad,block,serial,longitude,latitude,reused_count,outcome,OrbPay,launch_date
36,2026-079,Falcon 9,Low Earth Orbit,Space Launch Complex 40,7,True,LZ-40,Block 5,B1094,-80.577357,28.561941,6,True RTLS,12.000,2026-04-11
37,2026-077,Falcon 9,Low Earth Orbit,Space Launch Complex 4E,33,True,OCISLY,Block 5,B1063,-120.611000,34.632000,31,True ASDS,15.500,2026-04-11
40,2026-068,Falcon 9,Low Earth Orbit,Space Launch Complex 40,36,True,JRTI,Block 5,B1067,-80.577357,28.561941,33,True ASDS,16.100,2026-03-30
41,2026-067,Falcon 9,Sun-Synchronous Orbit,Space Launch Complex 4E,15,True,OCISLY,Block 5,B1093,-120.611000,34.632000,11,True ASDS,5.000,2026-03-30
46,2026-054,Falcon 9,Low Earth Orbit,Space Launch Complex 40,13,True,ASOG,Block 5,B1090,-80.577357,28.561941,10,True ASDS,16.100,2026-03-17
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
442,2023-037,Falcon 9,Low Earth Orbit,Space Launch Complex 4E,35,True,OCISLY,Block 5,B1071,-120.611000,34.632000,7,True ASDS,16.060,2023-03-17
462,2022-174,Falcon 9,Medium Earth Orbit,Space Launch Complex 40,36,True,ASOG,Block 5,B1067,-80.577357,28.561941,7,True ASDS,3.400,2022-12-16
463,2022-173,Falcon 9,Low Earth Orbit,Space Launch Complex 4E,35,True,LZ-4,Block 5,B1071,-120.611000,34.632000,5,True RTLS,2.200,2022-12-16
472,2022-125,Falcon 9,Low Earth Orbit,Space Launch Complex 4E,35,True,OCISLY,Block 5,B1071,-120.611000,34.632000,4,True ASDS,16.060,2022-10-05


So, it must be possible to still have a one-to-one inner join between the merged df and course df. But, first, launches with overlapping dates need to be dropped.

In [49]:
print('OVERALL DATE RANGE:\n merged_df: between {} and {}\n course_df: between {} and {}'.format(
    merged_df['launch_date'].min().year, merged_df['launch_date'].max().year,
    course_df['launch_date'].min().year, course_df['launch_date'].max().year
))

OVERALL DATE RANGE:
 merged_df: between 2010 and 2026
 course_df: between 2006 and 2020


In [50]:
merged_df = merged_df.drop(index=duplicate_dates.index)

In [51]:
try:
    merged_df = merged_df.merge(course_df, on='launch_date', how='inner', validate='1:1')
except Exception as e:
    print(e)

In [52]:
merged_df

,launch_designator,booster_version,orbit,launch_site,flights,reused,landing_pad,block,serial,longitude,latitude,reused_count,outcome,OrbPay,launch_date,gridfins,legs
0,2020-084,Falcon 9,Low Earth Orbit,Launch Complex 39A,23,False,JRTI,Block 5,B1061,-80.604282,28.608227,0,True ASDS,13.000,2020-11-16,True,True
1,2020-078,Falcon 9,Medium Earth Orbit,Space Launch Complex 40,23,False,OCISLY,Block 5,B1062,-80.577357,28.561941,0,True ASDS,4.400,2020-11-05,True,True
2,2020-074,Falcon 9,Low Earth Orbit,Space Launch Complex 40,20,True,JRTI,Block 5,B1060,-80.577357,28.561941,2,True ASDS,16.000,2020-10-24,True,True
3,2020-073,Falcon 9,Low Earth Orbit,Launch Complex 39A,14,True,OCISLY,Block 5,B1051,-80.604282,28.608227,5,True ASDS,16.000,2020-10-18,True,True
4,2020-070,Falcon 9,Low Earth Orbit,Launch Complex 39A,19,True,OCISLY,Block 5,B1058,-80.604282,28.608227,2,True ASDS,16.000,2020-10-06,True,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
65,2015-001,Falcon 9,Low Earth Orbit,Space Launch Complex 40,1,False,JRTI,v1.1,B1012,-80.577357,28.561941,0,False ASDS,12.410,2015-01-10,True,True
66,2014-056,Falcon 9,Low Earth Orbit,Space Launch Complex 40,1,False,ATL,v1.1,B1010,-80.577357,28.561941,0,False Ocean,12.303,2014-09-21,False,False
67,2013-055,Falcon 9,Low Earth Orbit,Space Launch Complex 4E,1,False,PAC,v1.1,B1003,-120.611000,34.632000,0,False Ocean,0.578,2013-09-29,False,False
68,2010-066,Falcon 9,Low Earth Orbit,Space Launch Complex 40,1,False,ATL,v1.0,B0004,-80.577357,28.561941,0,False PCL,8.774,2010-12-08,False,False


Finally, to modularize the merge

In [53]:
def merge_launch_data(ll2_df: pd.DataFrame, gcat_df: pd.DataFrame, course_df: pd.DataFrame, csv_path: None|str|Path = None) -> None|Path:
    '''
    Function to merge transformed ll2 launches data, transformed GCAT launch data, and transformed course data.
    ll2 launches and gcat data will be merged on `launch_designator`, and resulting merged DataFrame will be merged
    with course data on `launch_date`. Each merge is a one-to-one inner join. The final merged DataFrame will be
    saved to `csv_path`

    :param ll2_df: pd.DataFrame; transformed ll2 data
    :param gcat_df: pd.DataFrame; transformed gcat data
    :param course_df: pd.DataFrame; transformed course data
    :param csv_path: None or str or Path; Default None. File path where merged DataFrame is to be saved. If None,
                     merged DataFrame will be saved in the current working directory
    :return: Path; csv_path
    '''

    if not csv_path:
        csv_path = Path.cwd() / 'launch_data_1.csv'

    if isinstance(csv_path, str):
        csv_path = Path(csv_path)

    # Merge LL2 and GCAT
    merged_df = ll2_df.merge(gcat_df, on='launch_designator', how='inner', validate='1:1')

    mask = (merged_df['launch_date_x'] == merged_df['launch_date_y'])
    if mask.all():
        merged_df['launch_date'] = merged_df['launch_date_x']
        merged_df = merged_df.drop(['launch_date_x', 'launch_date_y'], axis=1)
    else:
        raise Exception('Mismatch: launch_date in LL2 Launches does not match with launch_date in GCAT Launches')
        return None

    # Drop launches with overlapping dates
    counts = merged_df['launch_date'].value_counts()
    duplicate_dates = merged_df[merged_df['launch_date'].map(lambda x: x in counts[counts>1].index)]
    merged_df = merged_df.drop(index=duplicate_dates.index)

    # Merge Course DF
    merged_df = merged_df.merge(course_df, on='launch_date', how='inner', validate='1:1')

    # Convert DF to CSV
    merged_df.to_csv(csv_path, index=False)

    return csv_path

In [54]:
csv_path = Path.cwd().parent / 'data' / 'interim' / 'api-launch-data-table.csv'
merge_launch_data(ll2_df, gcat_df, course_df, csv_path=csv_path)

PosixPath('/Users/jelo/spacey_falcon_9_project/data/interim/api-launch-data-table.csv')

## Trying out the modularized notebook

In [56]:
import spacey_falcon_9_project.dataset

In [57]:
# LL2 Launches
ll2_launches_file_paths = download_all_ll2_launches()

all_ll2_launches_path = merge_ll2_launch_data(
    ll2_launches_file_paths,
    Path.cwd().parent / 'data' / 'interim' / 'll2-api-2.3.0-launches-previous-merged.json'
)

ll2_df = transform_ll2_launches(all_ll2_launches_path)

# GCAT Launches
url_gcat='https://planet4589.org/space/gcat/tsv/launch/Falcon9.tsv'
file_name_gcat = 'mcdowell-gcat-launch-data.tsv'

gcat_file_path = download_launch_data_static(url_gcat, file_name_gcat)

gcat_df = transform_gcat_data(gcat_file_path)

# Course-provided data
url_ibm='https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/API_call_spacex_api.json'
file_name_ibm = 'ibm-ds-capstone-launch-data.json'

course_file_path = download_launch_data_static(url_ibm, file_name_ibm)

course_df = transform_course_data(course_file_path)

# Merge
merged_df = merge_launch_data(ll2_df, gcat_df, course_df,
                             Path.cwd().parent / 'data' / 'interim' / 'launch_data_1.csv')

ll2-api-2.3.0-launches-previous-0.json already exists.
ll2-api-2.3.0-launches-previous-100.json already exists.
ll2-api-2.3.0-launches-previous-200.json already exists.
ll2-api-2.3.0-launches-previous-300.json already exists.
ll2-api-2.3.0-launches-previous-400.json already exists.
ll2-api-2.3.0-launches-previous-500.json already exists.
ll2-api-2.3.0-launches-previous-600.json already exists.
mcdowell-gcat-launch-data.tsv already exists
ibm-ds-capstone-launch-data.json already exists


In [59]:
pd.read_csv(merged_df)

,launch_designator,booster_version,orbit,launch_site,flights,reused,landing_pad,block,serial,longitude,latitude,reused_count,outcome,OrbPay,launch_date,gridfins,legs
0,2020-084,Falcon 9,Low Earth Orbit,Launch Complex 39A,23,False,JRTI,Block 5,B1061,-80.604282,28.608227,0,True ASDS,13.000,2020-11-16,True,True
1,2020-078,Falcon 9,Medium Earth Orbit,Space Launch Complex 40,23,False,OCISLY,Block 5,B1062,-80.577357,28.561941,0,True ASDS,4.400,2020-11-05,True,True
2,2020-074,Falcon 9,Low Earth Orbit,Space Launch Complex 40,20,True,JRTI,Block 5,B1060,-80.577357,28.561941,2,True ASDS,16.000,2020-10-24,True,True
3,2020-073,Falcon 9,Low Earth Orbit,Launch Complex 39A,14,True,OCISLY,Block 5,B1051,-80.604282,28.608227,5,True ASDS,16.000,2020-10-18,True,True
4,2020-070,Falcon 9,Low Earth Orbit,Launch Complex 39A,19,True,OCISLY,Block 5,B1058,-80.604282,28.608227,2,True ASDS,16.000,2020-10-06,True,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
65,2015-001,Falcon 9,Low Earth Orbit,Space Launch Complex 40,1,False,JRTI,v1.1,B1012,-80.577357,28.561941,0,False ASDS,12.410,2015-01-10,True,True
66,2014-056,Falcon 9,Low Earth Orbit,Space Launch Complex 40,1,False,ATL,v1.1,B1010,-80.577357,28.561941,0,False Ocean,12.303,2014-09-21,False,False
67,2013-055,Falcon 9,Low Earth Orbit,Space Launch Complex 4E,1,False,PAC,v1.1,B1003,-120.611000,34.632000,0,False Ocean,0.578,2013-09-29,False,False
68,2010-066,Falcon 9,Low Earth Orbit,Space Launch Complex 40,1,False,ATL,v1.0,B0004,-80.577357,28.561941,0,False PCL,8.774,2010-12-08,False,False
